### PR TITLE
Make host uniqueness modular

### LINKED ISSUE #781: Allow HOST ID override
**Is your feature request related to a problem? Please describe.**
Some environments won't require the same level of scrutiny we apply to unique hosts.

Currently we need to write a file which we can't do reliably without elevated permissions on all OSes.

**Describe the solution you'd like**

Create an override module that evaluates the environment and returns a seeded UUID4. The module should serve as an example to others how they can implement it but will take an environment variable.

If no environment variable var is available should fail over to the file on disk. If an env var is present the on disk file location shouldn't be created.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by hulto**
#### What type of PR is this?
Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
This PR provides greater control to admins when deploying the imix agent around how the uniqueness of a host is determined. In most red teaming situations writing a file to disk is unideal and the tradeoffs to use something else like mac address are worth it. This PR adds an environment variable override to the host ID and makes it easier for admins to implement and add their own uniqueness.

Fixes #781

### CODE REVIEW

File: docs/_docs/user-guide/imix.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ Imix has compile-time configuration, that may be specified using environment var
 | IMIX_CALLBACK_INTERVAL | Duration between callbacks, in seconds. | `5` | No |
 | IMIX_RETRY_INTERVAL | Duration to wait before restarting the agent loop if an error occurs, in seconds. | `5` | No |
 | IMIX_PROXY_URI | Overide system settings for proxy URI over HTTP(S) (must specify a scheme, e.g. `https://`) | No proxy | No |
+| IMIX_HOST_ID | Manually specify the host ID for this beacon. Superceeds the file on disk. | - | No |
```
Comment (Cictrone) [inline docs/_docs/user-guide/imix.md:22]: supersedes

File: docs/_docs/dev-guide/imix.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+- Create an implementation of the `HostUniqueEngine`
+
+```rust
+use uuid::Uuid;
+
+use crate::HostUniqueEngine;
+
+pub struct MacAddress {}
+
+impl HostUniqueEngine for MacAddress {
```
Comment (Cictrone) [inline docs/_docs/dev-guide/imix.md:45]: HostIDSelector

File: implants/imix/src/config.rs [hunk 3]
```diff
@@ -78,9 +73,11 @@ impl Default for Config {
             identifier: format!("imix-v{}", VERSION),
         };
 
+        let engines = host_unique::defaults();
```
Comment (Cictrone) [inline implants/imix/src/config.rs:76]: i think its default not defaults

File: implants/imix/src/config.rs [hunk 4]
```diff
@@ -78,9 +73,11 @@ impl Default for Config {
...
+
         let host = pb::c2::Host {
             name: whoami::fallible::hostname().unwrap_or(String::from("")),
-            identifier: get_host_id(get_host_id_path()),
+            identifier: host_unique::id(engines).to_string(),
```
Comment (Cictrone) [inline implants/imix/src/config.rs:80]: get_id_with_selectors

File: implants/lib/host_unique/src/env.rs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+impl Default for Env {
+    fn default() -> Self {
+        Env {}
+    }
+}
+
+impl HostUniqueEngine for Env {
+    fn get_name(&self) -> String {
+        "env".to_string()
```
Comment (Cictrone) [inline implants/lib/host_unique/src/env.rs:16]: Maybe add more context?

**Review by Cictrone — CHANGES_REQUESTED** [review]

File: implants/imix/src/config.rs (hunk 3 above)
Comment (hulto) [reply implants/imix/src/config.rs:76]: Can't have default on a trait.

**Review by Cictrone — APPROVED** [review]

**Pull request merged**